File: AN_2/SEM_2/MPP/c#/lab5/Networking/mpp.networking/rpcprotocol/ClientJsonWorker.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using Model.mpp.model;
using Services.mpp.services;
using Exception = System.Exception;

namespace Networking.mpp.networking.rpcprotocol
{
    public class ClientJsonWorker : IObserver
    {
        private IServices server;
        private TcpClient connection;
        private NetworkStream stream;
        private volatile bool connected;

        public ClientJsonWorker(IServices server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                connected = true;
            }
            catch (IOException e)
            {
                Console.WriteLine("Stream initialization error: " + e.StackTrace);
            }
        }

        public void Run()
        {
            using StreamReader reader = new StreamReader(stream);
            while (connected)
            {
                try
                {
                    string requestJson = reader.ReadLine();
                    if (string.IsNullOrEmpty(requestJson)) continue;
                    Request request = JsonSerializer.Deserialize<Request>(requestJson);
                    Response response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (IOException e)
                {
                    Console.WriteLine("IOException: " + e.StackTrace);
                }
                Thread.Sleep(1000);
            }

            try
            {
                stream.Close();
                connection.Close();
            }
            catch (IOException e)
            {
                Console.WriteLine("Error closing resources: " + e.StackTrace);
            }
        }

        private void SendResponse(Response response)
        {
            string json = JsonSerializer.Serialize(response);
            lock (stream)
            {
                byte[] data = Encoding.UTF8.GetBytes(json + "\n");
                stream.Write(data, 0, data.Length);
                stream.Flush();
            }
        }


        public void Adauga(ParticipantCursa participantCursa)
        {
            Response response = JsonProtocolUtils.CreateNewParticipantCursaResponse(participantCursa);
            try
            {
                SendResponse(response);
            }
            catch (IOException e)
            {
                throw new Exception("Sending error: " + e);
            }
        }

        private static Response okResponse = JsonProtocolUtils.CreateOkResponse();

        private Response HandleRequest(Request request)
        {
            Response response = null;

            switch (request.type)
            {
                case RequestType.LOGIN:
                    response = HandleLoginRequest(request);
                    break;
                case RequestType.LOGOUT:
                    response = HandleLogoutRequest(request);
                    break;
                case RequestType.GET_ALL_CURSE:
                    response = HandleGetAllCurseRequest();
                    break;
                case RequestType.FIND_ECHIPA_BY_NUME:
                    response = HandleFindEchipaByNumeRequest(request);
                    break;
                case RequestType.GET_ALL_PARTICIPANTI:
                    response = HandleGetAllParticipantiRequest();
                    break;
                case RequestType.LIST_DTO_CURSE: 
                    response = HandleListDtoCurseRequest(request);
                    break;
                case RequestType.CAUTA_PARTICIPANTI_ECHIPA:
                    response = HandleCautaParticipantiEchipaRequest(request);
                    break;
                case RequestType.ADAUGA_ECHIPA:
                    response = HandleAdaugaEchipaRequest(request);
                    break;
                case RequestType.FIND_NR_PART_BY_CURSA:
                    response = HandleFindNrPartByCursaRequest(request);
                    break;
                case RequestType.GET_ALL_PARTICIPANT_CURSA:
                    response = HandleGetAllParticipantCursaRequest();
                    break;
                case RequestType.ADAUGA_PARTICIPANT_LA_CURSA:
                    response = HandleAdaugaParticipantLaCursaRequest(request);
                    break;
                case RequestType.INSCRIERE_PARTICIPANT:
                    response = HandleInscriereParticipantRequest(request);
                    break;
                case RequestType.FIND_PARTICIPANT_BY_NUME:
                    response = HandleFindParticipantByNumeRequest(request);
                    break;
                default:
                    response = JsonProtocolUtils.CreateErrorResponse("Unknown request type.");
                    break;
            }

            return response;
        }

        private Response HandleLoginRequest(Request request)
        {
            Utilizator utilizator = request.utilizator;

            if (utilizator == null || utilizator.Nume == null || utilizator.Parola == null)
                return JsonProtocolUtils.CreateErrorResponse("Login failed: utilizator invalid.");

            try
            {
                lock (server)
                    server.Login(utilizator, this);
                return okResponse;
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Login failed: " + e.Message);
            }
        }
        
        private Response HandleFindParticipantByNumeRequest(Request request)
        {
            string nume = request.numeParticipant;
            if (string.IsNullOrEmpty(nume))
                return JsonProtocolUtils.CreateErrorResponse("Nume participant lipsă.");

            try
            {
                Participant participant;
                lock (server)
                {
                    participant = server.FindByNume(nume);
                }

                if (participant == null)
                    return JsonProtocolUtils.CreateErrorResponse("Participantul nu a fost găsit.");

                return JsonProtocolUtils.CreateFindParticipantByNumeResponse(participant);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Eroare căutare participant: " + e.Message);
            }
        }

        
        private Response HandleFindEchipaByNumeRequest(Request request)
        {
            try
            {
                var echipa = server.FindEchipaByNume(request.numeEchipa);
                return JsonProtocolUtils.CreateFindEchipaByNumeResponse(echipa);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Eroare căutare echipă: " + e.Message);
            }
        }
        
        private Response HandleListDtoCurseRequest(Request request)
        {
            try
            {
                List<CursaDTO> curseDTOs;
                lock (server)
                {
                    curseDTOs = server.ListDtoCurse();
                }

                return JsonProtocolUtils.CreateListDtoCurseResponse(curseDTOs);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error handling LIST_DTO_CURSE request: " + e.Message);
            }
        }
        private Response HandleGetAllCurseRequest()
        {
            try
            {
                List<Cursa> curse;
                lock (server)
                {
                    curse = server.FindAllCurse();
                }

                return JsonProtocolUtils.CreateGetAllCurseResponse(curse);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error retrieving courses: " + e.Message);
            }
        }

        private Response HandleGetAllParticipantiRequest()
        {
            try
            {
                List<Participant> participanti;
                lock (server)
                {
                    participanti = server.FindAllParticipanti();
                }

                return JsonProtocolUtils.CreateGetAllParticipantiResponse(participanti);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error retrieving participants: " + e.Message);
            }
        }

        private Response HandleCautaParticipantiEchipaRequest(Request request)
        {
            string echipa = request.numeEchipa;

            if (echipa == null)
            {
                return JsonProtocolUtils.CreateErrorResponse("Echipa name is null.");
            }

            try
            {
                List<Participant> participanti;
                lock (server)
                {
                    participanti = server.CautaParticipantiDupaEchipa(echipa);
                }

                return JsonProtocolUtils.CreateCautaParticipantiEchipaResponse(participanti);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error searching participants: " + e.Message);
            }
        }

        private Response HandleAdaugaEchipaRequest(Request request)
        {
            Echipa echipa = request.echipa;

            if (echipa == null)
            {
                return JsonProtocolUtils.CreateErrorResponse("Invalid Echipa data.");
            }

            try
            {
                lock (server)
                    server.AdaugaEchipa(echipa);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error adding team: " + e.Message);
            }
        }
        
        

        private Response HandleFindNrPartByCursaRequest(Request request)
        {
            Cursa cursa = request.cursa;

            if (cursa == null)
            {
                return JsonProtocolUtils.CreateErrorResponse("Invalid Cursa data.");
            }

            try
            {
                int nrParticipanti;
                lock (server)
                {
                    nrParticipanti = server.FindNrPartByCursa(cursa);
                }

                return JsonProtocolUtils.CreateFindNrPartByCursaResponse(nrParticipanti);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error retrieving number of participants: " + e.Message);
            }
        }

        private Response HandleGetAllParticipantCursaRequest()
        {
            try
            {
                List<ParticipantCursa> participantCurse;
                lock (server)
                {
                    participantCurse = server.FindAllParticipantCursa();
                }

                return JsonProtocolUtils.CreateGetAllParticipantCursaResponse(participantCurse);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error retrieving participants for race: " + e.Message);
            }
        }
        
        private Response HandleLogoutRequest(Request request)
        {
            try
            {
                lock (server)
                {
                    server.Logout(request.utilizator, this);
                }

                connected = false; 
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Logout failed: " + e.Message);
            }
        }



        private Response HandleAdaugaParticipantLaCursaRequest(Request request)
        {

            ParticipantCursa participantCurse = request.participantCursa;

            if (participantCurse == null)
            {
                return JsonProtocolUtils.CreateErrorResponse("Invalid participant or race data.");
            }

            try
            {
                lock (server)
                    server.AdaugaParticipantLaCursa(participantCurse);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Error adding participant to race: " + e.Message);
            }
        }
        
        private Response HandleInscriereParticipantRequest(Request request)
        {
            Participant participant = request.participant;
            if (participant == null)
            {
                return JsonProtocolUtils.CreateErrorResponse("Invalid participant data.");
            }
            try
            {
                lock (server)
                {
                    server.InscriereParticipant(participant);
                }
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse("Inscriere failed: " + e.Message);
            }
        }
    }
}

File: AN_2/SEM_2/MPP/java/lab8-9/Networking/src/main/java/ro/mpp2024/network/jsonprotocol/ClientJsonWorker.java
package ro.mpp2024.network.jsonprotocol;

import com.google.gson.*;
import ro.mpp2024.model.*;
import ro.mpp2024.services.IObserver;
import ro.mpp2024.services.IServices;

import java.io.*;
import java.net.Socket;
import java.util.List;

public class ClientJsonWorker implements Runnable, IObserver {
    private final IServices server;
    private final Socket connection;
    private BufferedReader reader;
    private BufferedWriter writer;
    private final Gson gson = new GsonBuilder().create();
    private volatile boolean connected;

    public ClientJsonWorker(IServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            writer = new BufferedWriter(new OutputStreamWriter(connection.getOutputStream()));
            connected = true;
        } catch (IOException e) {
            System.err.println("Error initializing streams: " + e.getMessage());
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                String requestLine = reader.readLine();
                if (requestLine == null || requestLine.isEmpty()) continue;
                Request request = gson.fromJson(requestLine, Request.class);
                //System.out.println("Am primit: " + request);
                Response response = handleRequest(request);
                //System.out.println("Am trimis: " + response);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                System.err.println("IOException: " + e.getMessage());
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                System.err.println("Interrupted: " + e.getMessage());
            }
        }
        try {
            reader.close();
            writer.close();
            connection.close();
        } catch (IOException e) {
            System.err.println("Error closing resources: " + e.getMessage());
        }
    }

    private static final Response okResponse = JsonProtocolUtils.createOkResponse();

    private void sendResponse(Response response) throws IOException {
        String responseJson = gson.toJson(response);
        synchronized (writer) {
            writer.write(responseJson);
            writer.newLine();
            writer.flush();
        }
    }

    private Response handleRequest(Request request) {
        if (request == null || request.getType() == null) {
            return JsonProtocolUtils.createErrorResponse("Invalid request");
        }
        switch (request.getType()) {
            case LOGIN:
                return handleLogin(request);
            case LOGOUT:
                return handleLogout(request);
            case GET_ALL_CURSE:
                return handleGetAllCurse();
            case LIST_DTO_CURSE:
                return handleListDtoCurse();
            case GET_ALL_PARTICIPANTI:
                return handleGetAllParticipanti();
            case CAUTA_PARTICIPANTI_ECHIPA:
                return handleCautaParticipantiEchipa(request);
            case FIND_ECHIPA_BY_NUME:
                return handleFindEchipaByNume(request);
            case ADAUGA_ECHIPA:
                return handleAdaugaEchipa(request);
            case FIND_NR_PART_BY_CURSA:
                return handleFindNrPartByCursa(request);
            case GET_ALL_PARTICIPANT_CURSA:
                return handleGetAllParticipantCursa();
            case INSCRIERE_PARTICIPANT:
                return handleInscriereParticipant(request);
            case ADAUGA_PARTICIPANT_LA_CURSA:
                return handleAdaugaParticipantLaCursa(request);
            case FIND_PARTICIPANT_BY_NUME:
                return handleFindParticipantByNume(request);
            default:
                return JsonProtocolUtils.createErrorResponse("Unknown request type");
        }
    }

    private Response handleLogin(Request request) {
        Utilizator utilizator = request.getUtilizator();
        try {
            server.login(utilizator, this);
            return okResponse;
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse("Login failed: " + e.getMessage());
        }
    }

    private Response handleLogout(Request request) {
        try {
            server.logout(request.getUtilizator(), this);
            connected = false;
            return okResponse;
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse("Logout failed: " + e.getMessage());
        }
    }

    private Response handleGetAllCurse() {
        try {
            List<Cursa> curse = server.findAllCurse();
            return JsonProtocolUtils.createGetAllCurseResponse(curse);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleListDtoCurse() {
        try {
            List<CursaDTO> dtoList = server.listDtoCurse();
            return JsonProtocolUtils.createListDtoCurseResponse(dtoList);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleGetAllParticipanti() {
        try {
            List<Participant> participanti = server.findAllParticipanti();
            return JsonProtocolUtils.createGetAllParticipantiResponse(participanti);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleCautaParticipantiEchipa(Request request) {
        String echipa = request.getNumeEchipa();
        try {
            List<Participant> participanti = server.cautaParticipantiDupaEchipa(echipa);
            return JsonProtocolUtils.createCautaParticipantiEchipaResponse(participanti);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleFindEchipaByNume(Request request) {
        try {
            Echipa echipa = server.findEchipaByNume(request.getNumeEchipa());
            return JsonProtocolUtils.createFindEchipaByNumeResponse(echipa);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleAdaugaEchipa(Request request) {
        try {
            server.adaugaEchipa(request.getEchipa());
            return okResponse;
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleFindNrPartByCursa(Request request) {
        try {
            int nr = server.findNrPartByCursa(request.getCursa());
            return JsonProtocolUtils.createFindNrPartByCursaResponse(nr);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleGetAllParticipantCursa() {
        try {
            List<ParticipantCursa> list = server.findAllParticipantCursa();
            return JsonProtocolUtils.createGetAllParticipantCursaResponse(list);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleInscriereParticipant(Request request) {
        try {
            server.inscriereParticipant(request.getParticipant());
            return okResponse;
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleAdaugaParticipantLaCursa(Request request) {
        try {
            server.adaugaParticipantLaCursa(request.getParticipantCursa());
            return okResponse;
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    private Response handleFindParticipantByNume(Request request) {
        try {
            Participant participant = server.findByNume(request.getNumeParticipant());
            if (participant == null)
                return JsonProtocolUtils.createErrorResponse("Participantul nu a fost găsit.");
            return JsonProtocolUtils.createFindParticipantByNumeResponse(participant);
        } catch (Exception e) {
            return JsonProtocolUtils.createErrorResponse(e.getMessage());
        }
    }

    @Override
    public void adauga(Participant participant, Cursa cursa) throws Exception {
        ParticipantCursa pc = new ParticipantCursa(participant, cursa);
        Response response = JsonProtocolUtils.createNewParticipantCursaResponse(pc);
        sendResponse(response);
    }
}
